hpx/config/lambda_capture.hpp

```diff
@@ -0,0 +1,21 @@
+//  Copyright (c) 2018 Hartmut Kaiser
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef HPX_CONFIG_LAMBDA_CAPTURE_HPP
+#define HPX_CONFIG_LAMBDA_CAPTURE_HPP
+
+#include <hpx/config/defines.hpp>
+
+#include <utility>
+
+#if defined(HPX_HAVE_CXX14_LAMBDAS)
+#define HPX_CAPTURE_FORWARD(var, type)  var = std::forward<type>(var)
+#define HPX_CAPTURE_MOVE(var)           var = std::move(var)
+#else
+#define HPX_CAPTURE_FORWARD(var, type)  var
+#define HPX_CAPTURE_MOVE(var)           var
```

Comment:
Hmm, I don't think this is a good idea in general. We'll introduce inefficiencies when not having the captures or compile errors (if the objects are not copyable in C++11 mode, for example).

Reply:
Huh? This improves the picture for C++14 and leaves things unchanged as they are in C++11.

Reply:
I didn't through all changes. It might lead to those problems. It does improve the picture for C++14 indeed, but leaves C++11 code behind. As said, I'm not sure if it's a good idea. If we need to keep C++11 compatibility, a better option might be to capture via `bind`/`deferred_call`.

Reply:
Improving the picture for C++11 is independent of this PR.

Reply:
That's exactly why I'm saying this might not be a good idea. A solution that would work for both would be significantly different, and requires way more boilerplate. What you propose though, is a C++14 only solution (with a C++11 facade, that's not doing anything). I don't think it's bad to have it that way, but I think it should be without that macro that's not giving us any benefit for older compilers, and will lead to problems eventually.

Reply:
Sure, I understand. This is completely orthogonal to this PR, though. Things do not change for C++11 based compilation, no matter whether we apply this PR or not. But things improve for C++14 based compilation, so it's worth going ahead.

Reply:
> Sure, I understand. This is completely orthogonal to this PR, though. Things do not change for C++11 based compilation, no matter whether we apply this PR or not. 

Right, which I think is a potential problem ...

> But things improve for C++14 based compilation, so it's worth going ahead.

... as only C++14 code improves. there isn't any problem in particular right now. The problem is the code that will be written against those interfaces. We won't notice any problem since all our compilers actually support lambda init captures. There is no gain in artificially introducing this distinction, that's my main problem. We either solve the problem for all supported flavors of c++ properly (which I strongly support by bumping the requirements) or we keep it as it is to avoid future problems with the two incompatible code paths.

Reply:
@sithhell so you're refusing to accept an improvement for 80% of the systems we're running on for the sake of a potential problem which might accidentally be introduced by some careless change on 20% of the systems?

What alternative solution do you suggest?

Reply:
What I'm suggesting is to leave the 20% behind properly, by removing support for them and bump the minimal version to C++11 with C++14 lambda init captures. This way, we ensure that no such careless change sneaks in. Besides, problems will most likely manifest in user code first where we don't have any control.

Reply:
That would currently leave behind all of CUDA. I think that's not what we want.